cmd/kube-apiserver/app/server.go

```diff
@@ -126,6 +127,7 @@ cluster's shared state through which all other components interact.`,
 			return nil
 		},
 	}
+	cmd.SetContext(genericapiserver.SetupSignalContext())
```

Comment:
🤔  – some go proverb: don't store contexts.

Reply:
I know, but sometimes the alternative is breaking the API badly. I bet that was the reason for cobra's decision to add a command context. As that is now part of their API, we should use it.

Reply:
I understand why they do it. It's always about late initialization of a context that is wired in constructor code. So you have to cut the link and that only work by accessing a stored context.

Reply:
I need to check how Cobra handles sub-commands.

Regarding `genericapiserver.SetupSignalContext` that can be called more than once because it uses `signal.Notify` under the hood and that can be called more than once:
> It is allowed to call Notify multiple times with different channels
> and the same signals: each channel receives copies of incoming signals independently.

Reply:
Oh, wait:
https://github.com/kubernetes/apiserver/blob/9b2e61f60b8a7d50ccfb40adad37780cf4038cbf/pkg/server/signal.go#L41

Why? That seems unnecessary.

Reply:
> is the context from the root command automatically propagated to the sub commands?

Yes. I checked with this patch:
```patch
diff --git a/cmd/kubectl/kubectl.go b/cmd/kubectl/kubectl.go
index 09c18cfa209..5143366839a 100644
--- a/cmd/kubectl/kubectl.go
+++ b/cmd/kubectl/kubectl.go
@@ -21,12 +21,15 @@ import (
        "k8s.io/kubectl/pkg/cmd"
        "k8s.io/kubectl/pkg/cmd/util"
 
+       genericapiserver "k8s.io/apiserver/pkg/server"
+
        // Import to initialize client auth plugins.
        _ "k8s.io/client-go/plugin/pkg/client/auth"
 )
 
 func main() {
        command := cmd.NewDefaultKubectlCommand()
+       command.SetContext(genericapiserver.SetupSignalContext())
        if err := cli.RunNoErrOutput(command); err != nil {
                // Pretty-print the error and exit with an error.
                util.CheckErr(err)
diff --git a/staging/src/k8s.io/apiserver/pkg/server/signal.go b/staging/src/k8s.io/apiserver/pkg/server/signal.go
index e5334ae4c15..24883d02bfe 100644
--- a/staging/src/k8s.io/apiserver/pkg/server/signal.go
+++ b/staging/src/k8s.io/apiserver/pkg/server/signal.go
@@ -51,7 +51,7 @@ func SetupSignalContext() context.Context {
                os.Exit(1) // second signal. Exit directly.
        }()
 
-       return ctx
+       return context.WithValue(ctx, "foo", "bar")
 }
 
 // RequestShutdown emulates a received event that is considered as shutdown signal (SIGTERM/SIGINT)
```

Then in a sub-command like https://github.com/kubernetes/kubernetes/blob/5fe1e923df1fed37f437241d8d63808121fc3670/staging/src/k8s.io/kubectl/pkg/cmd/run/run.go#L157 `dlv` confirms that `cmd.ctx` is indeed the context prepared by `SetupSignalContext`:
```
> k8s.io/kubectl/pkg/cmd/run.NewCmdRun.func1() ./staging/src/k8s.io/kubectl/pkg/cmd/run/run.go:157 (hits goroutine(1):1 total:1) (PC: 0x3890707)
...
(dlv) p cmd.ctx
context.Context(*context.valueCtx) *{
	Context: context.Context(*context.cancelCtx) *{
		Context: context.Context(context.backgroundCtx) *(*context.Context)(0xc000522550),
		mu: (*sync.Mutex)(0xc000522560),
		done: (*"sync/atomic.Value")(0xc000522568),
		children: map[context.canceler]struct {} nil,
		err: error nil,
		cause: error nil,},
	key: interface {}(string) "foo",
	val: interface {}(string) "bar",}
```